File: NoTomorrow/Features/Progress/MeasurementsViews.swift
import SwiftUI
import SwiftData

/// Tape measurements: conversion (cm stored, inches shown to lb users), labels and the per-kind summary.
enum Measurements {
    static let cmPerInch = 2.54

    static func usesInches(_ kind: MeasurementKind, unit: WeightUnit) -> Bool { kind != .bodyFat && unit == .lb }

    static func display(_ value: Double, kind: MeasurementKind, unit: WeightUnit) -> Double {
        usesInches(kind, unit: unit) ? value / cmPerInch : value
    }

    static func stored(_ value: Double, kind: MeasurementKind, unit: WeightUnit) -> Double {
        usesInches(kind, unit: unit) ? value * cmPerInch : value
    }

    static func unitLabel(_ kind: MeasurementKind, unit: WeightUnit) -> String {
        kind == .bodyFat ? "%" : (unit == .lb ? "in" : "cm")
    }

    /// "84,5 cm", "33.3 in", "18,5 %".
    static func label(_ value: Double, kind: MeasurementKind, unit: WeightUnit, signed: Bool = false) -> String {
        let shown = display(value, kind: kind, unit: unit)
        let number = shown.formatted(.number.precision(.fractionLength(0...1))
            .sign(strategy: signed ? .always(includingZero: false) : .automatic).locale(Fmt.locale))
        return "\(number)\u{00A0}\(unitLabel(kind, unit: unit))"
    }

    static func titleKey(_ kind: MeasurementKind) -> LocalizedStringKey {
        LocalizedStringKey("measure." + kind.rawValue)
    }

    struct Reading: Equatable {
        var kind: MeasurementKind
        var day: Date
        var value: Double
    }

    struct Summary: Equatable {
        var kind: MeasurementKind
        var latest: Reading
        /// Latest minus the first reading, when there are two or more.
        var change: Double?
    }

    /// One summary per kind that has readings, in `MeasurementKind` order.
    static func summaries(_ readings: [Reading]) -> [Summary] {
        MeasurementKind.allCases.compactMap { kind in
            let mine = readings.filter { $0.kind == kind }.sorted { $0.day < $1.day }
            guard let first = mine.first, let last = mine.last else { return nil }
            return Summary(kind: kind, latest: last, change: mine.count >= 2 ? last.value - first.value : nil)
        }
    }

    /// Today's value per kind is replaced; an empty field leaves that kind alone.
    @MainActor
    static func save(_ values: [MeasurementKind: Double], day: Date = .now, in context: ModelContext) {
        let start = Calendar.current.startOfDay(for: day)
        let existing = (try? context.fetch(FetchDescriptor<BodyMeasurement>(predicate: #Predicate { $0.day == start }))) ?? []
        for (kind, value) in values where value > 0 {
            if let row = existing.first(where: { $0.kindRaw == kind.rawValue }) {
                row.value = value
            } else {
                context.insert(BodyMeasurement(day: start, kind: kind, value: value))
            }
        }
        try? context.save()
    }
}

// MARK: - Section

/// Progress > Body, under the weight card: the latest of each measurement and its change since the first one,
/// and Log measurements.
struct MeasurementsSection: View {
    var unit: WeightUnit

    @Query(sort: \BodyMeasurement.day) private var rows: [BodyMeasurement]
    @State private var showsLog = false

    private var readings: [Measurements.Reading] {
        rows.compactMap { row in row.kind.map { Measurements.Reading(kind: $0, day: row.day, value: row.value) } }
    }

    var body: some View {
        let summaries = Measurements.summaries(readings)
        VStack(alignment: .leading, spacing: 0) {
            SectionHeader(title: "measure.title").padding(.bottom, 4)
            if summaries.isEmpty {
                Text("measure.empty")
                    .font(NT.Fonts.subheadline).foregroundStyle(NT.Colors.ink2)
                    .fixedSize(horizontal: false, vertical: true)
                    .padding(.vertical, 12)
            } else {
                ForEach(Array(summaries.enumerated()), id: \.offset) { index, summary in
                    if index > 0 { Hairline() }
                    row(summary)
                }
            }
            SecondaryButton(title: "measure.log", systemImage: "ruler", height: NT.Size.cardButton) { showsLog = true }
                .padding(.top, 12)
        }
        .sheet(isPresented: $showsLog) {
            LogMeasurementsSheet(unit: unit, last: Dictionary(summaries.map { ($0.kind, $0.latest.value) }, uniquingKeysWith: { a, _ in a }))
        }
    }

    private func row(_ summary: Measurements.Summary) -> some View {
        HStack(spacing: 8) {
            VStack(alignment: .leading, spacing: 2) {
                Text(Measurements.titleKey(summary.kind)).font(NT.Fonts.subheadline).foregroundStyle(NT.Colors.ink)
                Text(Fmt.dayMonth(summary.latest.day)).font(NT.Fonts.caption).foregroundStyle(NT.Colors.ink2)
            }
            Spacer()
            if let change = summary.change, abs(change) >= 0.05 {
                Text(Measurements.label(change, kind: summary.kind, unit: unit, signed: true))
                    .font(NT.Fonts.footnote).foregroundStyle(NT.Colors.ink2).tabular()
            }
            Text(Measurements.label(summary.latest.value, kind: summary.kind, unit: unit))
                .font(NT.Fonts.headline).foregroundStyle(NT.Colors.ink).tabular()
                .frame(minWidth: 84, alignment: .trailing)
        }
        .frame(minHeight: 52)
        .accessibilityElement(children: .combine)
    }
}

// MARK: - Log sheet

/// One field per measurement, last values as placeholders; Save writes today's readings for the filled fields.
struct LogMeasurementsSheet: View {
    var unit: WeightUnit
    /// Latest stored value per kind (cm / %), for the placeholders.
    var last: [MeasurementKind: Double]

    @Environment(\.modelContext) private var modelContext
    @Environment(\.dismiss) private var dismiss
    @State private var texts: [MeasurementKind: String] = [:]

    private var parsed: [MeasurementKind: Double] {
        var result: [MeasurementKind: Double] = [:]
        for (kind, text) in texts {
            let value = SetInput.number(text)
            guard value > 0, value < (kind == .bodyFat ? 80 : 400) else { continue }
            result[kind] = Measurements.stored(value, kind: kind, unit: unit)
        }
        return result
    }

    var body: some View {
        VStack(spacing: 0) {
            HStack {
                Button { dismiss() } label: {
                    Text("common.cancel").font(NT.Fonts.body).foregroundStyle(NT.Colors.ink2).frame(minHeight: NT.Size.control)
                }
                Spacer()
                Text("measure.log").font(NT.Fonts.headline).foregroundStyle(NT.Colors.ink)
                Spacer()
                Button {
                    Measurements.save(parsed, in: modelContext)
                    Haptics.success()
                    dismiss()
                } label: {
                    Text("common.save").font(NT.Fonts.headline)
                        .foregroundStyle(parsed.isEmpty ? NT.Colors.ink3 : NT.Colors.ember)
                        .frame(minHeight: NT.Size.control)
                }
                .disabled(parsed.isEmpty)
            }
            .padding(.horizontal, NT.Spacing.screenH)
            .padding(.top, 16)
            ScrollView {
                VStack(alignment: .leading, spacing: 12) {
                    Text(Fmt.longDay(.now)).font(NT.Fonts.footnote).foregroundStyle(NT.Colors.ink2)
                    STGroup {
                        ForEach(MeasurementKind.allCases, id: \.self) { kind in
                            field(kind)
                        }
                    }
                }
                .padding(.horizontal, NT.Spacing.screenH)
                .padding(.top, 12)
                .padding(.bottom, NT.Spacing.section)
            }
            .scrollDismissesKeyboard(.interactively)
        }
        .ntScreenBackground()
        .presentationBackground(NT.Colors.ground)
        .presentationDragIndicator(.visible)
    }

    private func field(_ kind: MeasurementKind) -> some View {
        let placeholder = last[kind].map {
            Measurements.display($0, kind: kind, unit: unit)
                .formatted(.number.precision(.fractionLength(0...1)).locale(Fmt.locale))
        } ?? "—"
        return HStack(spacing: 12) {
            Text(Measurements.titleKey(kind)).font(NT.Fonts.body).foregroundStyle(NT.Colors.ink)
                .accessibilityHidden(true)
            Spacer(minLength: 8)
            TextField("", text: Binding(get: { texts[kind] ?? "" }, set: { texts[kind] = $0 }),
                      prompt: Text(verbatim: placeholder).foregroundStyle(NT.Colors.ink3))
                .keyboardType(.decimalPad)
                .multilineTextAlignment(.trailing)
                .font(NT.Fonts.body).foregroundStyle(NT.Colors.ink).tabular()
                .frame(width: 90)
                .accessibilityLabel(Text(Measurements.titleKey(kind))
                    + Text(verbatim: ", " + Measurements.unitLabel(kind, unit: unit)))
            Text(verbatim: Measurements.unitLabel(kind, unit: unit))
                .font(NT.Fonts.subheadline).foregroundStyle(NT.Colors.ink2)
                .fixedSize()
                .frame(minWidth: 24, alignment: .leading)
                .accessibilityHidden(true)
        }
        .frame(minHeight: NT.Size.control)
    }
}

File: android/app/src/main/java/app/notomorrow/feature/progress/MeasurementsViews.kt
package app.notomorrow.feature.progress

import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.heightIn
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.layout.widthIn
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.text.BasicTextField
import androidx.compose.foundation.text.KeyboardOptions
import androidx.compose.foundation.verticalScroll
import androidx.compose.runtime.Composable
import androidx.compose.runtime.collectAsState
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateMapOf
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.SolidColor
import androidx.compose.ui.hapticfeedback.HapticFeedbackType
import androidx.compose.ui.platform.LocalHapticFeedback
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.semantics.clearAndSetSemantics
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.text.input.KeyboardType
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.dp
import app.notomorrow.designsystem.Hairline
import app.notomorrow.designsystem.NT
import app.notomorrow.designsystem.NtIcons
import app.notomorrow.designsystem.NtSheet
import app.notomorrow.designsystem.NtShapes
import app.notomorrow.designsystem.NtText
import app.notomorrow.designsystem.SecondaryButton
import app.notomorrow.designsystem.SectionHeader
import app.notomorrow.designsystem.TabularText
import app.notomorrow.designsystem.ntDismissKeyboardOnScroll
import app.notomorrow.designsystem.ntPlainClickable
import app.notomorrow.designsystem.tabular
import app.notomorrow.di.LocalAppContainer
import app.notomorrow.model.MeasurementKind
import app.notomorrow.model.WeightUnit
import app.notomorrow.service.Days
import app.notomorrow.util.Fmt
import app.notomorrow.util.NtKeys
import app.notomorrow.util.S
import java.time.LocalDate
import kotlin.math.abs
import kotlinx.coroutines.launch

/**
 * Progress › Body, under the weight card — `MeasurementsSection`: the latest of each measurement
 * and its change since the first one, and Log measurements.
 */
@Composable
fun MeasurementsSection(unit: WeightUnit, modifier: Modifier = Modifier) {
    val dao = LocalAppContainer.current.db.bodyMeasurementDao()
    val rows by remember(dao) { dao.observeAll() }.collectAsState(initial = emptyList())
    val summaries = Measurements.summaries(Measurements.readings(rows))
    var showsLog by remember { mutableStateOf(false) }

    Column(modifier.fillMaxWidth()) {
        SectionHeader(title = stringResource(S.measure_title), modifier = Modifier.padding(bottom = 4.dp))
        if (summaries.isEmpty()) {
            NtText(
                text = stringResource(S.measure_empty),
                modifier = Modifier.padding(vertical = 12.dp),
                style = NT.Fonts.subheadline,
                color = NT.Colors.ink2,
            )
        } else {
            summaries.forEachIndexed { index, summary ->
                if (index > 0) Hairline()
                SummaryRow(summary, unit)
            }
        }
        SecondaryButton(
            title = stringResource(S.measure_log),
            modifier = Modifier.padding(top = 12.dp),
            icon = NtIcons.Plus,
            height = NT.Size.cardButton,
            onClick = { showsLog = true },
        )
    }

    if (showsLog) {
        LogMeasurementsSheet(
            unit = unit,
            last = summaries.associate { it.kind to it.latest.value },
            onDismiss = { showsLog = false },
        )
    }
}

@Composable
private fun SummaryRow(summary: Measurements.Summary, unit: WeightUnit) {
    Row(
        modifier = Modifier.fillMaxWidth().heightIn(min = 52.dp),
        horizontalArrangement = Arrangement.spacedBy(8.dp),
        verticalAlignment = Alignment.CenterVertically,
    ) {
        Column(Modifier.weight(1f), verticalArrangement = Arrangement.spacedBy(2.dp)) {
            NtText(stringResource(NtKeys.measure(summary.kind)), style = NT.Fonts.subheadline, color = NT.Colors.ink)
            NtText(
                text = Fmt.dayMonth(Days.date(summary.latest.day)),
                style = NT.Fonts.caption,
                color = NT.Colors.ink2,
            )
        }
        val change = summary.change
        if (change != null && abs(change) >= 0.05) {
            TabularText(
                text = Measurements.label(change, summary.kind, unit, signed = true),
                style = NT.Fonts.footnote,
                color = NT.Colors.ink2,
            )
        }
        NtText(
            text = Measurements.label(summary.latest.value, summary.kind, unit),
            modifier = Modifier.widthIn(min = 84.dp),
            style = NT.Fonts.headline.tabular(),
            color = NT.Colors.ink,
            maxLines = 1,
            textAlign = TextAlign.End,
        )
    }
}

/**
 * `LogMeasurementsSheet` — one field per measurement, last values as placeholders; Save writes
 * today's readings for the filled fields. [last] is the latest stored value per kind (cm / %).
 */
@Composable
private fun LogMeasurementsSheet(
    unit: WeightUnit,
    last: Map<MeasurementKind, Double>,
    onDismiss: () -> Unit,
) {
    val dao = LocalAppContainer.current.db.bodyMeasurementDao()
    val scope = rememberCoroutineScope()
    val haptics = LocalHapticFeedback.current
    val texts = remember { mutableStateMapOf<MeasurementKind, String>() }
    val parsed = Measurements.parsed(texts, unit)

    NtSheet(onDismiss = onDismiss, showsHandle = true, containerColor = NT.Colors.ground) {
        Row(
            modifier = Modifier
                .fillMaxWidth()
                .padding(horizontal = NT.Spacing.screenH)
                .padding(top = 16.dp),
            verticalAlignment = Alignment.CenterVertically,
        ) {
            Box(
                modifier = Modifier.heightIn(min = NT.Size.control).ntPlainClickable(onClick = onDismiss),
                contentAlignment = Alignment.Center,
            ) {
                NtText(stringResource(S.common_cancel), style = NT.Fonts.body, color = NT.Colors.ink2, maxLines = 1)
            }
            NtText(
                text = stringResource(S.measure_log),
                modifier = Modifier.weight(1f),
                style = NT.Fonts.headline,
                color = NT.Colors.ink,
                maxLines = 1,
                textAlign = TextAlign.Center,
            )
            Box(
                modifier = Modifier
                    .heightIn(min = NT.Size.control)
                    .ntPlainClickable(enabled = parsed.isNotEmpty()) {
                        val values = parsed
                        scope.launch {
                            Measurements.save(values, dao)
                            haptics.performHapticFeedback(HapticFeedbackType.Confirm)
                            onDismiss()
                        }
                    },
                contentAlignment = Alignment.Center,
            ) {
                NtText(
                    text = stringResource(S.common_save),
                    style = NT.Fonts.headline,
                    color = if (parsed.isEmpty()) NT.Colors.ink3 else NT.Colors.ember,
                    maxLines = 1,
                )
            }
        }
        Column(
            modifier = Modifier
                .fillMaxWidth()
                .ntDismissKeyboardOnScroll()
                .verticalScroll(rememberScrollState())
                .padding(horizontal = NT.Spacing.screenH)
                .padding(top = 12.dp, bottom = NT.Spacing.section),
            verticalArrangement = Arrangement.spacedBy(12.dp),
        ) {
            NtText(Fmt.longDay(LocalDate.now()), style = NT.Fonts.footnote, color = NT.Colors.ink2)
            Column(
                Modifier
                    .fillMaxWidth()
                    .background(NT.Colors.surface, NtShapes.tile)
                    .padding(horizontal = 16.dp),
            ) {
                MeasurementKind.entries.forEachIndexed { index, kind ->
                    if (index > 0) Hairline()
                    MeasurementField(
                        kind = kind,
                        unit = unit,
                        text = texts[kind].orEmpty(),
                        last = last[kind],
                        onText = { texts[kind] = it },
                    )
                }
            }
        }
    }
}

@Composable
private fun MeasurementField(
    kind: MeasurementKind,
    unit: WeightUnit,
    text: String,
    last: Double?,
    onText: (String) -> Unit,
) {
    val placeholder = last?.let { Measurements.number(Measurements.display(it, kind, unit)) } ?: "—"
    Row(
        modifier = Modifier.fillMaxWidth().heightIn(min = NT.Size.control),
        horizontalArrangement = Arrangement.spacedBy(12.dp),
        verticalAlignment = Alignment.CenterVertically,
    ) {
        val title = stringResource(NtKeys.measure(kind))
        val fieldLabel = title + ", " + Measurements.unitLabel(kind, unit)
        NtText(title, modifier = Modifier.clearAndSetSemantics {}, style = NT.Fonts.body, color = NT.Colors.ink, maxLines = 1)
        Spacer(Modifier.weight(1f).widthIn(min = 8.dp))
        Box(Modifier.width(90.dp), contentAlignment = Alignment.CenterEnd) {
            if (text.isEmpty()) {
                NtText(placeholder, style = NT.Fonts.body.tabular(), color = NT.Colors.ink3, maxLines = 1)
            }
            BasicTextField(
                value = text,
                onValueChange = onText,
                modifier = Modifier.fillMaxWidth().semantics { contentDescription = fieldLabel },
                textStyle = NT.Fonts.body.tabular().copy(color = NT.Colors.ink, textAlign = TextAlign.End),
                singleLine = true,
                cursorBrush = SolidColor(NT.Colors.ink),
                keyboardOptions = KeyboardOptions(keyboardType = KeyboardType.Decimal),
            )
        }
        NtText(
            text = Measurements.unitLabel(kind, unit),
            modifier = Modifier.widthIn(min = 24.dp).clearAndSetSemantics {},
            style = NT.Fonts.subheadline,
            color = NT.Colors.ink2,
            maxLines = 1,
        )
    }
}
